<b>

<p>
<center>
<font size="6">
SC1015 Mini-Project
</font>
</center>
</p>

<p>
<center>
<font size="5">
Drug Performance Evaluation: Predicting Drug Efficacy
</font>
</center>
</p>

<p>
<center>
<font size="3">
SC1015: Data Science and Artifial Intelligence, Nanyang Technological University
</font>
</center>
</p>

<p>
<center>
<font size="3">
Done By: Chan Kit Ho and Chiang Qin Zhi
</font>
</center>
</p>

</b>

In [162]:
# Basic Libraries
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt # we only need pyplot
sb.set() # set the default Seaborn style for graphics

In [163]:
DrugData = pd.read_csv('Drug.csv')
DrugData.head()


Condition          Drug Indication Type      Reviews  \
0  Acute Bacterial Sinusitis  Levofloxacin   On Label   RX  994 Reviews   
1  Acute Bacterial Sinusitis  Levofloxacin   On Label   RX  994 Reviews   
2  Acute Bacterial Sinusitis  Moxifloxacin   On Label   RX  755 Reviews   
3  Acute Bacterial Sinusitis  Azithromycin   On Label   RX  584 Reviews   
4  Acute Bacterial Sinusitis  Azithromycin   On Label   RX  584 Reviews   

   Effective  EaseOfUse  Satisfaction  \
0       2.52       3.01          1.84   
1       2.52       3.01          1.84   
2       2.78       3.00          2.08   
3       3.21       4.01          2.57   
4       3.21       4.01          2.57   

                                         Information  
0  \n\t\t\t\t\tLevofloxacin is used to treat a va...  
1  \n\t\t\t\t\tLevofloxacin is used to treat a va...  
2  \n\t\t\t\t\t This is a generic drug. The avera...  
3  \n\t\t\t\t\tAzithromycin is an antibiotic (mac...  
4  \n\t\t\t\t\tAzithromycin is an antibiotic (mac...

In [164]:
DrugData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2219 entries, 0 to 2218
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Condition     2219 non-null   object 
 1   Drug          2219 non-null   object 
 2   Indication    2219 non-null   object 
 3   Type          2219 non-null   object 
 4   Reviews       2219 non-null   object 
 5   Effective     2219 non-null   float64
 6   EaseOfUse     2219 non-null   float64
 7   Satisfaction  2219 non-null   float64
 8   Information   2219 non-null   object 
dtypes: float64(3), object(6)
memory usage: 156.1+ KB


Let's list down all our observations from the super-quick glance of the dataset, as above.

There are 9 variables/features/columns and 2219 observations/samples/rows in the dataset.

The response variable seems to be Effectiveness, while the remaining 8 are most likely predictors.

There are 3 variables identified as float64 by default, and it seems they are indeed Numeric.

There are 6 variables identified as object by default, and they are most likely Categorical (except Reviews,which seems to be numerical count).

None of the variables/features seem to have any missing value.

---
## Data Preparation and Cleaning

We will look through the dataset to prepare and indentify any areas that needs to be cleaned before our analysis.

In [165]:
DrugData["Reviews"]

0       994 Reviews
1       994 Reviews
2       755 Reviews
3       584 Reviews
4       584 Reviews
           ...     
2214      2 Reviews
2215      1 Reviews
2216      1 Reviews
2217      1 Reviews
2218      1 Reviews
Name: Reviews, Length: 2219, dtype: object

Looking at the Reviews Variable, we note that Reviews should be Numerical but it is classified as object64 due to the string behind. Hence, there is a need to remove "reviews" and convert it to a numerical variable.

In [166]:

DrugData["Reviews"] = DrugData["Reviews"].map(lambda x: x.rstrip(' Reviews')).astype('int64')
DrugData["Reviews"]


0       994
1       994
2       755
3       584
4       584
       ... 
2214      2
2215      1
2216      1
2217      1
2218      1
Name: Reviews, Length: 2219, dtype: int64

The price and form of medicine can be found within the information column. We will then extract the price and form from the column and create new columns for price and form. (Example of forms are "tablet", "capsule" etc.)

In [167]:
all_form = ['Capsule(s)','Capsule','Tablet(s)','Tablet','tablets','Bottle','Vial(s)','Vial','Reconstituted(s)','Reconstituted','Tube','Jar','Can','Box','Syringe','Implant','Package','Pen(s)','Inhaler']
Form = []
Price = []
for i in range (DrugData.shape[0]):
    info = DrugData.iloc[i,8]
    sentence = info.split('. ')
    for x in sentence:
        if 'average' in x :
            words = x.split()
    for word in words:
        word = word.replace(",", "")
        if word in all_form :
            Form.append(word)
            break
    for price in words:
        if '$' in price:
            temp = price.replace("$", "")
            Price.append(temp)

In [168]:
tablet = ['Tablet(s)','Tablet','tablets']
capsule = ['Capsule(s)','Capsule']
cream = ['Tube','Can','Jar']
liquid_drink = ['Bottle']
liquid_inject = ['Vial','Reconstituted','Reconstituted(s)','Vial(s)','Pen(s)','Syringe']
other = ['Box','Package','Implant','Inhaler']
for i in range (len(Form)):
    if Form[i] in tablet:
        Form[i] = 'Tablet'
    elif Form[i] in capsule:
        Form[i] = 'Capsule'
    elif Form[i] in cream:
        Form[i] = 'Cream'
    elif Form[i] in liquid_drink :
        Form[i] = 'Liquid (Drink)'
    elif Form[i] in liquid_inject :
        Form[i] = 'Liquid (Inject)'
    elif Form[i] in other:
        Form[i] = 'Other'

In [169]:
DrugData["Form"] = Form
DrugData["Price"] = Price

DrugData = DrugData.drop(['Information'], axis=1)
DrugData.head(10)
DrugData.shape


(2219, 10)

In the final cleaning process of the data, we want to ensure that we have accurate and non-duplicate information for each drug. To do this, we create a pivot table using the pd.pivot_table function with the index columns being the drug condition, drug name, indication, drug type, and drug form, and the aggregation function being the mean of the review ratings. 

This allows us to group drugs with the same name, form, and condition together and find the average review rating for each group. Finally, we create a new dataframe Data2 and loop through each index of the pivot table, appending the relevant information to Data2. This process ensures that we have accurate and non-duplicate drug review data, which can be used for further analysis.

In [170]:
#Change String to Float
DrugData['Reviews'] = DrugData['Reviews'].astype(float)
DrugData['Effective'] = DrugData['Effective'].astype(float)
DrugData['EaseOfUse'] = DrugData['EaseOfUse'].astype(float)
DrugData['Satisfaction'] = DrugData['Satisfaction'].astype(float)
DrugData['Price'] = DrugData['Price'].astype(float)

df = pd.pivot_table(DrugData, index=['Condition','Drug','Indication','Type','Form'],aggfunc='mean')
Data2 = pd.DataFrame()
for i in range (len(df.index)):
        temp = {'Condition': df.index[i][0],'Drug':df.index[i][1],'Indication':df.index[i][2],'Type':df.index[i][3],'Form':df.index[i][4],'EaseOfUse':df.iloc[i,0],'Effective':df.iloc[i,1] ,'Price':df.iloc[i,2],'Reviews':df.iloc[i,3] ,'Satisfaction':df.iloc[i,4]} 
        Data2 = Data2.append(temp, ignore_index=True)
Data2

/var/folders/5d/jp7tn7f54kd_q04htj05bhlm0000gn/T/ipykernel_20531/3928791598.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  Data2 = Data2.append(temp, ignore_index=True)
/var/folders/5d/jp7tn7f54kd_q04htj05bhlm0000gn/T/ipykernel_20531/3928791598.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  Data2 = Data2.append(temp, ignore_index=True)
/var/folders/5d/jp7tn7f54kd_q04htj05bhlm0000gn/T/ipykernel_20531/3928791598.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  Data2 = Data2.append(temp, ignore_index=True)
/var/folders/5d/jp7tn7f54kd_q04htj05bhlm0000gn/T/ipykernel_20531/3928791598.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat i

Condition                         Drug Indication  \
0    Acute Bacterial Sinusitis                  Amoxicillin   On Label   
1    Acute Bacterial Sinusitis  Amoxicillin-Pot Clavulanate  Off Label   
2    Acute Bacterial Sinusitis  Amoxicillin-Pot Clavulanate   On Label   
3    Acute Bacterial Sinusitis                   Ampicillin   On Label   
4    Acute Bacterial Sinusitis                   Ampicillin   On Label   
..                         ...                          ...        ...   
680   vulvovaginal candidiasis           Miconazole Nitrate   On Label   
681   vulvovaginal candidiasis      Miconazole-Skin Clnsr17   On Label   
682   vulvovaginal candidiasis      Miconazole-Skin Clnsr17   On Label   
683   vulvovaginal candidiasis                  Terconazole   On Label   
684   vulvovaginal candidiasis                  Tioconazole   On Label   

       Type            Form  EaseOfUse  Effective       Price     Reviews  \
0        RX         Capsule   3.852353   3.655882   12.590000   86.294118   
1        RX  Liquid (Drink)   3.470000   3.290000  287.370000   43.000000   
2        RX          Tablet   3.121429   2.962857   70.608571  267.285714   
3        RX         Capsule   2.000000   3.000000   12.590000    1.000000   
4        RX          Tablet   3.250000   3.000000  125.240000   15.000000   
..      ...             ...        ...        ...         ...         ...   
680  RX/OTC           Cream   3.465000   2.770000   13.990000   19.500000   
681     OTC           Cream   4.750000   3.000000   13.990000    4.000000   
682     OTC           Other   4.000000   1.000000  125.990000    1.000000   
683      RX           Cream   3.525000   3.047500   68.990000   20.000000   
684     OTC           Other   3.852500   2.022500   22.990000  145.000000   

     Satisfaction  
0        3.197647  
1        2.590000  
2        2.248571  
3        1.000000  
4        3.000000  
..            ...  
680      2.345000  
681      3.000000  
682      1.000000  
683      2.717500  
684      1.827500  

[685 rows x 10 columns]